### PR TITLE
PR for issue #44

### PR DESCRIPTION
**Pull request by chrismolina159**
Description:
The problem as described here #44 is solved by calling a serverless function `finishGame()` in `callFinishGame()` once the game decides that there are no more valid moves left.

What I changed:
I updated the database rules to allow the game to be ended at a time earlier than now(changed == to <="), and I added logic to make sure that even if the frontend makes repeated calls to update the game's status to **done**(due to `useEffect()` being called a few times when debugging), the program won't crash.

### CODE REVIEW

**Comment by ekzhang** [comment]
Thanks so much for fixing #44 @chrismolina159! I haven't been able to look at this until now, but I'm more available on weekends in general. Let me go through this.

File: database.rules.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
         },
         "endedAt": {
           ".write": "auth != null && data.parent().child('users/' + auth.uid).exists() && !data.exists() && newData.exists()",
```
Comment (ekzhang) [inline database.rules.json:11]: You can get rid of lines 10-13. The `endedAt` field does not need to be written from an external user, just from the trusted serverless admin API, which has full access anyway.

File: functions/src/index.ts [hunk 2]
```diff
@@ -6,6 +6,45 @@ admin.initializeApp();
...
+    .database()
+    .ref(`gameData/${gameId}/events`)
+    .limitToLast(1)
+    .once(
+      "value",
+      function (snapshot) {
+        for (const p in snapshot.val()) finalTime = snapshot.val()[p].time;
+      },
+      function (err) {
+        console.error(err);
```
Comment (ekzhang) [inline functions/src/index.ts:41]: Generally you should prefer to also throw a `functions.https.HttpsError()` in these cases. Also, as a note, this code might be cleaner using the return value of async/await (and a try/catch) instead of a callback, but I think the way you have it written is fine.

File: public/index.html [hunk 3]
```diff
@@ -27,10 +27,8 @@
   </head>
 
   <body>
-    <noscript>
-      You need to enable JavaScript to run this app.
```
Comment (ekzhang) [inline public/index.html:31]: Why is this changed? We're automatically formatting this project using Prettier. Please run `npm run format`, and it will be consistent across the codebase.

File: src/pages/GamePage.js [hunk 4]
```diff
@@ -142,6 +139,29 @@ function GamePage({ match }) {
...
+  async function callFinishGame() {
+    let timeDataRef = await finishGame({ gameId: gameId });
+    let endTime = timeDataRef.data.endedAt;
+    if (game.status === "ingame") {
+      firebase
+        .database()
+        .ref(`games/${gameId}`)
+        .update(
+          {
+            status: "done",
```
Comment (ekzhang) [inline src/pages/GamePage.js:151]: The point of the serverless function was so it can update the game-ended time from a trusted source. As you have it currently set up, users can essentially set the endedAt time to an arbitrary value in the past, which is insecure.

You should have this database update inside the serverless function itself.

File: src/pages/GamePage.js [hunk 5]
```diff
@@ -142,6 +139,29 @@ function GamePage({ match }) {
...
+            endedAt: endTime,
+          },
+          (err) => {
+            if (err) {
+              console.log("Already updated the game's status");
```
Comment (ekzhang) [inline src/pages/GamePage.js:156]: Remove debug print statements before merging

**Review by ekzhang — CHANGES_REQUESTED** [review]
This looks like great progress. I think the most important thing is that from a conceptual standpoint, we want to update the database from the serverless function itself, not from the untrusted client code. This prevents users from doing adversarial things to the `endedAt` value.

File: public/index.html (hunk 3 above)
Comment (chrismolina159) [reply public/index.html:31]: I did run `npm run format`, and it changed the above few lines to be on one line, so I reverted it back manually to stay consistent.

File: src/pages/GamePage.js (hunk 4 above)
Comment (chrismolina159) [reply src/pages/GamePage.js]: Should be more secure now

**Comment by chrismolina159** [comment]
Yeah don't worry if you take a second to go through this. I typically am the opposite and more available during the week. Let me know how you are feeling about the new commits/comments!

**Review by ekzhang — COMMENTED** [review]
Thanks so much @chrismolina159! This looks really great. I think there's just one remaining bug, which is that anyone can call `finishGame`, even when the game hasn't ended (or even started) yet! We need some amount of validation on the server side (in the function) to check if the game has actually finished, based on the events.

Doing this might require some validation similar to `computeState()` on the client side (see [here](https://github.com/ekzhang/setwithfriends/blob/99da67a9081a46a9d877b914e55a49d8daf2c6cb/src/util.js)), although it isn't necessary to be as picky about tracking scores and the ordering of cards. Do you think you could add a validation that there are actually no more sets in the game? This would require grabbing the entire `gameData` object, not just the last event. If there are remaining sets, then the function is being called improperly (either bug or bad actor) -> throw an HttpsError.

Thanks again. By the way, I'm going to merge upstream changes to this branch so you have the latest version.

**Comment by chrismolina159** [comment]
@ekzhang Glad to hear that it is looking good so far. Quick question in regards to your comment. The server function should check if the game exists, and the client should validate if there aren't any more sets correct?

I'm slightly confused on the client side part. How are you suggesting to go about this? In `useEffect()` the `finishGame` function only gets called if there aren't any sets available, but are you saying that someone can call the function mid-game to prematurely end the game? If so, would creating a flag based on `findSet`'s return be sufficient or what would you suggest? Thanks and hopefully we can get this done soon!

**Comment by ekzhang** [comment]
Client code is untrusted, as anyone can change it to do as they please. The server interface should be secure in itself, and a malicious (or more likely, unsynchronized) user should not be able to end the game while there are still sets remaining, or before the game has started.

I'll add some validation.

**Review by ekzhang — APPROVED** [review]

**Comment by ekzhang** [comment]
I just made some revisions, feel free to take a look and let me know what you think. @chrismolina159

**Comment by chrismolina159** [comment]
@ekzhang I think it looks good! Using `useRef()` and moving the functions onto the server to avoid repetition was a good idea.  Is the only thing left for us to do is have you merge the PR and close the issue?

**Comment by ekzhang** [comment]
Excellent, thanks so much for the help Chris!
